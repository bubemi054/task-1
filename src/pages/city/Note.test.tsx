// @ts-expect-error Need React imported for my tests
import React from "react";
import { render, cleanup, screen, fireEvent } from "@testing-library/react";
import NoteCard from "./Note";
import { describe, vi, it, expect, afterEach } from "vitest";
import type { Note } from "../../state-manager/types";
import "@testing-library/jest-dom/vitest";

afterEach(() => {
  cleanup();
  vi.restoreAllMocks();
});

describe("NoteCard Component", () => {
  const mockNote: Note = { id: 1, text: "Test Note", cityId: 123 };
  const mockOnSave = vi.fn();
  const mockOnDelete = vi.fn();

  it("renders the note text", () => {
    render(
      <NoteCard
        initialNote={mockNote}
        onSave={mockOnSave}
        onDelete={mockOnDelete}
      />,
    );
    expect(screen.getByDisplayValue(mockNote.text)).toBeInTheDocument();
  });

  it("enters edit mode when pencil icon is clicked", () => {
    render(
      <NoteCard
        initialNote={mockNote}
        onSave={mockOnSave}
        onDelete={mockOnDelete}
      />,
    );

    const editButton = screen.getByRole("edit-icon");
    fireEvent.click(editButton);

    const textarea = screen.getByRole("textbox");
    expect(textarea).not.toBeDisabled();
  });

  it("updates note text on change", () => {
    render(
      <NoteCard
        initialNote={mockNote}
        onSave={mockOnSave}
        onDelete={mockOnDelete}
      />,
    );

    const textarea = screen.getByRole("textbox");
    fireEvent.change(textarea, { target: { value: "Updated Note" } });

    expect(textarea).toHaveValue("Updated Note");
  });

  it("saves note when check icon is clicked", () => {
    render(
      <NoteCard
        initialNote={mockNote}
        onSave={mockOnSave}
        onDelete={mockOnDelete}
      />,
    );

    fireEvent.click(screen.getByRole("edit-icon"));

    const textarea = screen.getByRole("textbox");
    fireEvent.change(textarea, { target: { value: "Updated Note" } });

    fireEvent.click(screen.getByRole("save-icon"));

    expect(mockOnSave).toHaveBeenCalledWith({
      ...mockNote,
      text: "Updated Note",
    });
  });

  it("deletes note when trash icon is clicked", () => {
    render(
      <NoteCard
        initialNote={mockNote}
        onSave={mockOnSave}
        onDelete={mockOnDelete}
      />,
    );

    const deleteButton = screen.getByRole("delete-icon");
    fireEvent.click(deleteButton);

    expect(mockOnDelete).toHaveBeenCalledWith(mockNote);
  });

  it("toggles edit mode correctly", () => {
    render(
      <NoteCard
        initialNote={mockNote}
        onSave={mockOnSave}
        onDelete={mockOnDelete}
      />,
    );

    const editButton = screen.getByRole("edit-icon");

    fireEvent.click(editButton);
    expect(screen.queryByRole("save-icon")).toBeInTheDocument();
    expect(screen.queryByRole("edit-icon")).not.toBeInTheDocument();
    expect(screen.queryByRole("delete-icon")).not.toBeInTheDocument();

    fireEvent.click(screen.queryByRole("save-icon")!);
    expect(screen.queryByRole("edit-icon")).toBeInTheDocument();
    expect(screen.queryByRole("delete-icon")).toBeInTheDocument();
    expect(screen.queryByRole("save-icon")).toBeNull();
  });
});
