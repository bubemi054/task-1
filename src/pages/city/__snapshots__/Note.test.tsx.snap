// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NoteCard Component > matches snapshot 1`] = `
<div>
  <div
    class="relative"
  >
    <textarea
      class="flex h-[319px] w-[319px] flex-col justify-between rounded-lg bg-[rgba(0,0,0,0.2)] p-4 text-white backdrop-blur-[20px] outline-none [@media(max-width:380px)]:w-[100%]"
      disabled=""
    >
      Test Note
    </textarea>
    <div>
      <svg
        class="lucide lucide-trash absolute right-11 bottom-3 cursor-pointer text-white"
        fill="none"
        height="25"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M3 6h18"
        />
        <path
          d="M19 6v14c0 1-1 2-2 2H7c-1 0-2-1-2-2V6"
        />
        <path
          d="M8 6V4c0-1 1-2 2-2h4c1 0 2 1 2 2v2"
        />
      </svg>
      <svg
        class="lucide lucide-pencil absolute right-3 bottom-3 cursor-pointer text-white"
        fill="none"
        height="25"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="25"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z"
        />
        <path
          d="m15 5 4 4"
        />
      </svg>
    </div>
  </div>
</div>
`;
